src/coreclr/jit/optimizer.cpp

```diff
@@ -4493,77 +4533,51 @@ PhaseStatus Compiler::optUnrollLoops()
                 assert(totalIter > 0);
                 noway_assert(initBlock->KindIs(BBJ_ALWAYS));
             }
 
-            // The loop will be removed, so no need to fix up the pre-header.
-            if (loop.lpFlags & LPFLG_HAS_PREHEAD)
-            {
-                assert(head->HasFlag(BBF_LOOP_PREHEADER));
-
-                // For unrolled loops, all the unrolling preconditions require the pre-header block to fall
-                // through into TOP.
-                assert(head->KindIs(BBJ_ALWAYS));
-                assert(head->JumpsToNext());
-            }
-
-            // If we actually unrolled, tail is now reached
-            // by the last cloned bottom, and no longer
-            // reached by bottom.
-            //
-            if (totalIter > 0)
-            {
-                BasicBlock* clonedBottom = blockMap[bottom];
-                assert(clonedBottom->KindIs(BBJ_ALWAYS));
-                assert(!clonedBottom->HasInitializedTarget());
-
-                clonedBottom->SetTarget(tail);
-                fgAddRefPred(tail, clonedBottom);
-                fgRemoveRefPred(tail, bottom);
-            }
-
 #ifdef DEBUG
             if (verbose)
             {
                 printf("Whole unrolled loop:\n");
 
                 gtDispTree(initStmt->GetRootNode());
                 printf("\n");
-                fgDumpTrees(top, insertAfter);
-
-                if (anyNestedLoopsUnrolledThisLoop)
-                {
-                    printf("Unrolled loop " FMT_LP " contains nested loops\n", lnum);
-                }
+                fgDumpTrees(bottom, insertAfter);
             }
 #endif // DEBUG
 
+            // TODO: Remove
             // Update loop table.
-            optMarkLoopRemoved(lnum);
-
-            // Note if we created new BBJ_RETURNs (or removed some).
-            if (totalIter > 0)
-            {
-                fgReturnCount += loopRetCount * (totalIter - 1);
-            }
-            else
-            {
-                assert(totalIter == 0);
-                assert(fgReturnCount >= loopRetCount);
-                fgReturnCount -= loopRetCount;
-            }
+            optMarkLoopRemoved((unsigned)(oldLoop - optLoopTable));
 
             // Remember that something has changed.
             INDEBUG(++unrollCount);
             change = true;
+
+            for (FlowGraphNaturalLoop* ancestor = loop->GetParent(); ancestor != nullptr;
+                 ancestor                       = ancestor->GetParent())
+            {
+                BitVecOps::AddElemD(&loopTraits, loopsWithUnrolledDescendant, ancestor->GetIndex());
+            }
         }
 
     DONE_LOOP:;
     }
 
+    if (change && !BitVecOps::IsEmpty(&loopTraits, loopsWithUnrolledDescendant) && (passes < 10))
+    {
+        fgDomsComputed = false;
+        fgRenumberBlocks(); // For proper lexical visit
+        m_dfsTree = fgComputeDfs();
+        optFindNewLoops();
+        passes++;
+        goto RETRY_UNROLL;
```

Comment:
This pass could use some factoring into a few separate functions... I'll do that as a mechanical follow-up.